### PR TITLE
Adding support of handling complex masks applied to Layer group.

### PR DESCRIPTION
**Pull request by swap-null**
If fixes the bug : Quick Export Not Respecting Document Boundaries (file with layer mask) - 4019840

### CODE REVIEW

**Comment by redmunds** [comment]
reviewing

File: lib/renderer.js [hunk 1]
```diff
@@ -1134,6 +1134,27 @@
...
         return hasZeroBounds(this._getContentBounds(component.document.layers));
     };
+
+    /**
+     * checks to see if there is any complex mask applied to any of the layerGroup. 
+     * Complex Mask is defined as : 
+     * (1) Bitmap masks with extendWithWhite Property (Photoshop>Layers>Layer Mask>Hide Selection)
+     * (2) Vector masks or Bitmap Mask with zero bounds.(Layer Mask>Hide All or Vector Mask>Hide All)
+     * @private
+     * @param {!Component} component
```
Comment (redmunds) [inline lib/renderer.js]: There should also be an `@return` annotation. There doesn't seem to be one on the other 2 `_componentHas...` methods either, so can you update those also?

Comment (swap-null) [inline lib/renderer.js]: Done.

**Comment by swap-null** [comment]
@redmunds  once you finish the review, let me know. I will push the PR commits and then merge master into the branch.

File: lib/renderer.js [hunk 2]
```diff
@@ -1134,6 +1134,27 @@
...
+     * (1) Bitmap masks with extendWithWhite Property (Photoshop>Layers>Layer Mask>Hide Selection)
+     * (2) Vector masks or Bitmap Mask with zero bounds.(Layer Mask>Hide All or Vector Mask>Hide All)
+     * @private
+     * @param {!Component} component 
+     */
+    PixmapRenderer.prototype._componentHasComplexMasks = function (component) {
+        var layerGroupHasComplexMasks = function (layer) {
+            return layer.layers &&
+                    (layer.mask && layer.mask.enabled && layer.mask.extendWithWhite) ||
+                    ((layer.mask || layer.path) && !layer.getTotalMaskBounds());
```
Comment (redmunds) [inline lib/renderer.js]: Fix grouping (ask discussed in Slack)

**Comment by redmunds** [comment]
Done with code review

File: lib/renderer.js [hunk 3]
```diff
@@ -1134,6 +1136,28 @@
...
+     * (2) Vector masks or Bitmap Mask with zero bounds.(Layer Mask>Hide All or Vector Mask>Hide All)
+     * @private
+     * @param {!Component} component 
+     * @return {boolean}
+     */
+    PixmapRenderer.prototype._componentHasComplexMasks = function (component) {
+        var layerGroupHasComplexMasks = function (layer) {
+            return layer.layers &&
+                    ((layer.mask && layer.mask.enabled && layer.mask.extendWithWhite) ||
+                    ((layer.mask || layer.path) && !layer.getTotalMaskBounds()));
```
Comment (redmunds) [inline lib/renderer.js]: Style nit: can you add a space to beginning of last line to show grouping? For example:

```js
            return layer.layers &&
                   ((layer.mask && layer.mask.enabled && layer.mask.extendWithWhite) ||
                    ((layer.mask || layer.path) && !layer.getTotalMaskBounds()));
```

Comment (swap-null) [inline lib/renderer.js]: done.

**Comment by timothynoel** [comment]
A general comment about layer measuring:  The Photoshop compositing model is complex, can evolve over time. We might add new behavior or masks, and even the current behavior is hard to duplicate.  There's currently 3 kinds of masks, layer/vector/filter + the alpha mask for the layer content itself, there's vector strokes which can act like masks, fx, mask parameters like feather and density. A mask can apply the the layer "shape" or not, which affects the way fx are applied, and therefore affects the bounds of the layer. And there's tons of other tweaky things  like knockout, blend ranges, etc. that also affect the bounds.  (see Photoshop's Advanced Blending dialog for more info)

So it's probably better to leave any measuring logic inside Photoshop (beyond total layer/group bounds), or try to abstract this into something that Photoshop can provide in a robust/future proof way.

I realize that's not really practical right now, and you need the patch, just throwing it out there for future consideration.

**Comment by swap-null** [comment]
Merge was complex and there were risk of overwriting the changes. 
Created https://github.com/adobe-photoshop/generator-assets/pull/378 for the above changes. 
Closing this one.
@timothynoel: Thanks a lot for the insights. Totally agree to your points and will keep that for future consideration.

**Pull request closed without merging**